sources/startup.c

```diff
@@ -1773,22 +1776,74 @@ dontremove:;
 
 /*
  		#] CleanUp : 
- 		#[ Terminate :
+ 		#[ TerminateImpl :
 */
 
 static int firstterminate = 1;
 
-VOID Terminate(int errorcode)
+VOID TerminateImpl(int errorcode, const char* file, int line, const char* function)
 {
 	if ( errorcode && firstterminate ) {
 		firstterminate = 0;
+
+		MLOCK(ErrorMessageLock);
 #ifdef WITHPTHREADS
 		MesPrint("Program terminating in thread %w at &");
 #elif defined(WITHMPI)
 		MesPrint("Program terminating in process %w at &");
 #else
 		MesPrint("Program terminating at &");
 #endif
+		MesPrint("Terminate called from %s:%d (%s)", file, line, function);
+
+#ifdef ENABLE_BACKTRACE
+		void *stack[64];
+		int stacksize, stop = 0;
+		stacksize = backtrace(stack, sizeof(stack)/sizeof(stack[0]));
+
+		/* First check whether eu-addr2line is available */
+		if ( !system("command -v eu-addr2line > /dev/null 2>&1") ) {
+			MesPrint("Backtrace:");
+			for (int i = 0; i < stacksize && !stop; i++) {
+				FILE *fp;
+				char cmd[512];
+				MesPrint("%#%2d: %", i);
+				snprintf(cmd, sizeof(cmd)/sizeof(cmd[0]), "eu-addr2line -s --pretty-print -f -i '%p' --pid=%d\n", stack[i], getpid());
+				fp = popen(cmd, "r");
+				while ( fgets(cmd, sizeof(cmd), fp) != NULL ) {
+					MesPrint("%s", cmd);
+					/* Don't show functions lower than "main" */
+					if ( strstr(cmd, "main") || strstr(cmd, "RunThread") || strstr(cmd, "RunSortBot") ) {
```

Comment:
This condition potentially also stops at `BalanceRunThread` etc.??

Reply:
I guess `"main "`, `"RunThread "`, `"RunSortBot"` always appear at the beginning of the buffer, so a kind of `starts_with` would work.

Maybe `while` is not needed.

Reply:
I can check for " main " etc by leaving a space at cmd[0] and writing the result at `cmd+1`.

The while is required for multi-line output, for eg inlines (though we are suppressing this function usually)
```
#10: call_init at libc-start.c:128:20
  (inlined by) __libc_start_main_impl at libc-start.c:347:5
```

Reply:
Now I see the reason for `while`. Thanks.

And, yes checking " main " or (nonalphabetic character or start) + "main" + (nonalphabetic character) will work.